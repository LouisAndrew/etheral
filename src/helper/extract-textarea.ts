import { snakeCase } from 'lodash';

/**
 * function to extract options from a textarea
 * @param textarea textarea input (e.g sizes: S, M, L)
 */
export const extractTextArea = (textarea: string) =>
    textarea.split(',').map(text => snakeCase(text).toUpperCase());
