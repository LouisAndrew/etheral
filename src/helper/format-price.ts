export const formatPrice = (price: number | string) => {
    const value =
        typeof price === 'string'
            ? parseFloat(price.replace('IDR ', ''))
            : price;
    return new Intl.NumberFormat('id').format(value);
};
