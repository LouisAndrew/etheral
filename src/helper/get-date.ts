const pad2 = (n: number): string => (n < 10 ? `0${n}` : n.toString());

const getDate = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7);
    const minutes = pad2(date.getUTCMinutes());

    return year + '-' + month + '-' + day + ' ' + hours + ':' + minutes;
};

const getDateIpaymu = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7); // UTC offset for WIB => +7
    const minutes = pad2(date.getUTCMinutes());
    const seconds = pad2(date.getUTCSeconds());

    return `${year}${month}${day}${hours}${minutes}${seconds}`;
};

const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const getDateReadable = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = date.getUTCMonth();
    const day = date.getUTCDate();

    return `${day} ${monthNames[month]} ${year}`;
};

export { getDate, getDateIpaymu, getDateReadable };
