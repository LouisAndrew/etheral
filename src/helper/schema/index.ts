export { Blog } from './blog';
export { Collection } from './collection';
export { Product } from './product';
export { Order } from './order';
