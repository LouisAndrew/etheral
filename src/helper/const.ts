import { Product } from './schema/product';
import { Order } from './schema/order';

const mockProduct1: Product = {
    pid: '1234',
    name: 'Mock product',
    slug: 'ctg1/product1',
    category: 'ctg1',
    idrPrice: 23000,
    collection: 'fall',
    productImages: [],
    urls: [],
    weight: 20,
    amount: 12,
};

export const mockProducts: Product[] = [mockProduct1];

export const mockOrder: Order = {
    oid: 'asdadwda',
    buyerName: 'Jane Doe',
    buyerEmail: 'jane@doe',
    buyerPhone: 123412412,
    buyerUId: 'admin',
    buyerAddr: 'Storkowerstrasse',
    buyerPostal: 1234143,
    total: 32000,
    currency: 'IDR',
    date: new Date(),
    via: 'va',
    channel: 'bni',
    products: [{ pid: mockProduct1.pid, amount: 12 }],
    paid: false,
    delivered: false,
    transactionData: {
        sessionId: '1234',
        paymentNo: '12',
        paymentName: 'Jane Doe - BNI',
        expired: '1231241412',
        fee: 3000,
    },
    shippingMethod: 'OKE',
};

export const mockOrders: Order[] = [mockOrder];

export const mockOrderShipped: Order = {
    ...mockOrder,
    delivered: true,
    shippingData: {
        shippedDate: new Date(),
        trackingCode: '12342',
        shippedBy: 'admin1',
    },
};
