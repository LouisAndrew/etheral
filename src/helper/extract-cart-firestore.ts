import { intersectionBy } from 'lodash';

import { SetCartArgs } from 'state/actions/cart';
import { InCart } from './schema/firebase-user';
import { Product } from './schema/product';

export default ({
    firestoreCartData,
    allProducts,
}: {
    firestoreCartData: InCart;
    allProducts: Product[];
}): SetCartArgs => {
    const inCartProduct = intersectionBy(allProducts, firestoreCartData, 'pid');
    return inCartProduct.map(product => ({
        product,
        amount: intersectionBy(firestoreCartData, [product], 'pid')[0].amount, // get the intersection between the product -> pid is unique, so the function is guaranteed to return one item.
    }));
};
