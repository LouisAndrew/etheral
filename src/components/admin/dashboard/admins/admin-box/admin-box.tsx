import React from 'react';

import { Box, Text } from 'rebass';

import { Admin } from 'helper/schema';
import DeleteButton from 'components/admin/dashboard/delete-button';

type Props = Admin & {
    bg: string;
    adminEmail: string;
    removeAdmin: () => void;
};

const AdminBox: React.FC<Props> = ({
    bg,
    email,
    invitedBy,
    removeAdmin,
    adminEmail,
}) => {
    const textStyling = {
        fontFamily: 'body',
        fontSize: [1],
        // color: 'black.0' => removed to be able to apply custom color on remove text,
        fontWeight: 'medium',
        css: `
			 white-space: nowrap;
			overflow: hidden;
			text-overflow: ellipsis;
		`,
    };

    return (
        <Box
            bg={bg}
            sx={{
                display: 'grid',
                gridTemplateColumns: 'repeat(3, minmax(calc(256px / 3), 1fr))',
                gridGap: 2,
            }}
            p={[2]}
        >
            <Text {...textStyling} color="black.0">
                {email}
            </Text>
            <Text {...textStyling} color="black.0">
                {invitedBy}
            </Text>
            {invitedBy.toLowerCase() !== 'root' && adminEmail !== email ? (
                <DeleteButton onClick={removeAdmin} />
            ) : null}
        </Box>
    );
};

export { AdminBox };
