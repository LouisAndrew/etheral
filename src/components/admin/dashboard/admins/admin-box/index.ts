
import { AdminBox } from './admin-box';

export default AdminBox;