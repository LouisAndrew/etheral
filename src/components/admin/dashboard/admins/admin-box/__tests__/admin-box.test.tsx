import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import userEvent from '@testing-library/user-event';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { AdminBox } from '../admin-box';
import { Admin } from 'helper/schema';

const mockEmail = 'loha@gmail.com';

const mockAdmin: Admin = {
    email: 'lolo@mock',
    invitedBy: mockEmail,
    password: '1234',
};

describe('AdminBox', () => {
    const mockremoveAdmin = jest.fn(() => {});

    const Element = (
        <AdminBox
            bg=""
            adminEmail={mockEmail}
            removeAdmin={mockremoveAdmin}
            {...mockAdmin}
        />
    );
    const ElementSelf = (
        <AdminBox
            bg=""
            adminEmail={mockAdmin.email}
            removeAdmin={mockremoveAdmin}
            {...mockAdmin}
        />
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('should render the remove admin button  correctly', () => {
        const { queryByRole } = render(Element);

        const RemoveButton = queryByRole('button', { name: 'Remove' });

        if (!RemoveButton) {
            fail();
        }

        expect(RemoveButton).toBeInTheDocument();

        // test onClick
        userEvent.click(RemoveButton);
        expect(mockremoveAdmin).toBeCalled();
    });

    it('Should not render the remove button if user logged in has the same email with the element rednered', () => {
        const { queryByRole } = render(ElementSelf);

        const RemoveButton = queryByRole('button', { name: 'Remove' });

        if (!RemoveButton) {
            expect(true).toBe(true);
        }
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
