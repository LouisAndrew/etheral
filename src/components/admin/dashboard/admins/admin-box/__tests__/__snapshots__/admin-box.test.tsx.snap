// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminBox matches snapshot 1`] = `
<div
  className="css-1cx2s4a"
>
  <div
    className="css-welxbt"
  >
    lolo@mock
  </div>
  <div
    className="css-welxbt"
  >
    loha@gmail.com
  </div>
  <div
    className="css-vurnku"
  >
    <div
      className="css-1i8tx31"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [
              Class {
                "_dispatchInstances": null,
                "_dispatchListeners": null,
                "_targetInst": null,
                "altKey": null,
                "bubbles": null,
                "button": null,
                "buttons": null,
                "cancelable": null,
                "clientX": null,
                "clientY": null,
                "ctrlKey": null,
                "currentTarget": [Function],
                "defaultPrevented": null,
                "detail": null,
                "dispatchConfig": null,
                "eventPhase": null,
                "getModifierState": [Function],
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": null,
                "metaKey": null,
                "movementX": [Function],
                "movementY": [Function],
                "nativeEvent": null,
                "pageX": null,
                "pageY": null,
                "relatedTarget": [Function],
                "screenX": null,
                "screenY": null,
                "shiftKey": null,
                "target": null,
                "timeStamp": [Function],
                "type": null,
                "view": null,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      role="button"
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M17 6h5v2h-2v13a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1V8H2V6h5V3a1 1 0 0 1 1-1h8a1 1 0 0 1 1 1v3zm-8 5v6h2v-6H9zm4 0v6h2v-6h-2zM9 4v2h6V4H9z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
      Remove
    </div>
  </div>
</div>
`;
